Clarity and structure improvements in the shortcuts documentations (#786)
darktable-org/dtdocs: content/preferences-settings/shortcuts.md

draft: false
---

> _Shortcuts_ are a way to control darktable using your keyboard, pointing device or external controller. They allow you to perform _actions_ without interacting directly with a UI element.

Author: Done.

An _action_ is usually (but not always) an operation that you might undertake using darktable's point-and-click user interface.
See [common actions](#common-actions) below for a list of the most common ones. For example:

-   Increase, decrease or reset sliders
-   Scroll through dropdown lists
-   Enable, expand or focus modules
-   Click buttons
-   Switch between views

> 

Reviewer: Is this no longer the recommended way to assign shortcuts? This should be mentioned early on if it is still the recommended way.

Author: Done.

The simplest shortcut consists of just one key press (e.g., `E`).

The same key can be repeated up to three times, so also `E + E` and `E + E + E` are valid, and distinct, shortcuts.

You can further extend a shortcut with up to three clicks of different mouse buttons. So, `E + E + left-click` or `E + right-click + left-click` are also valid shortcuts.

Finally, if your fingers are very nimble and you really need to go there, you can repeat the last click up to three times. So, also `E + left-click + middle-click + middle-click` and `E + E + middle-click + left-click + right-click + right-click + right-click` are valid shortcuts.

Hence, the most complex shortcut can consist of 8 key/button presses:
* Three key presses of the same `key`, followed by
* Three presses of the three mouse buttons, followed by
* 2 more repetitions of the last click.

Your shortcut can include one or more modifiers (`Shift`, `Ctrl` and `Alt`). In this case, for a continuous shortcut, both the last key and the modifier must be pressed down when you execute the movement. For example, `Ctrl + E + E + pan` means that while moving the mouse left and/or right both  `E` and `Ctrl` must be pressed down.

---

> **Note:**

Reviewer: This note should probably go in the introductory section, if we include it at all.

If you are a MacOs user, your shortcuts will use `Cmd` instead of `Ctrl` and `Option` instead of `Alt`.

---

If you are defining a continuous shortcut, then the movement part of the shortcut must be executed while the last key, mouse or controller button is held down.

For example, the shortcut `E + E + pan`, can be activated by pressing `E` twice, holding down `E` on the second press and moving the mouse horizontally while the key is pressed.

**Short and long key presses.** 
By default, all key presses in a shortcut are _short_, i.e., the key is pressed and immediately released. However, the last repetition of a key in a shortcut can also be a _long_ key press, defined as holding down the key for a bit longer than the duration of a double click. Hence `E + E` and `E + E(long)` are two distinct shortcuts that can be assigned to different actions. The associated action triggers when the key is released, which entails that a shortcut ending with a long press cannot be used for a continuos shortcut.

**Triggering multiple shortcuts at once.** 

> If both `E + mouse-scroll` and `F + mouse-scroll` are mapped to a slider (or if you have fallbacks enabled), then scrolling the mouse wheel while holding down both `E` and `F` will move both sliders.

Reviewer: It would be nice to refer to the short cuts the same way as the preferences panel does, e.g. it doesn't call it "mouse scroll" just "scroll".

Similarly, you could map, say, `E + E + mouse-scroll` and `F + F + mouse-scroll` to two nodes in a graph (say, `yellow` and `green` in color equalizer). You can then use `E + E(hold) + F(hold) + mouse-scroll` to move both sliders at once. This works because the repeat key counter is per-shortcut, not per key, and it is reset only after the current shortcut ends (i.e., when the repeated key is released). Thus it is possible to trigger as many actions at once as your fingers can handle, provided that all the shortcuts involved have the same number of key presses (i.e., single, double or triple presses).

**Shortcuts must be unique within a view.**
A single action may have multiple shortcuts but a single shortcut can only be linked to one action in a given darktable view -- you can't chain actions together except by applying a preset or style. You can, however, set up a single shortcut that does one thing in the lighttable view, say, and another in the darkroom view.

> **Additional modifiers.** As mentioned above, the only valid modifiers are the `Shift`, `Ctrl` and `Alt` keys on the keyboard. You can define additional keys (or device buttons) as modifiers by assigning keys/buttons to the "global/modifier" action. However, these will merely function as extra `Ctrl`, `Alt` or `Shift` keys -- you cannot create "new" modifiers.

Reviewer: I find this paragraph confusing. So I can add new keys, but they'll function as ctrl/alt/shift, but I can't add new modifiers?

draft: false
---

> _Shortcuts_ are a way to control darktable using your keyboard, mouse or trackpad and other input devices. They allow you to perform _actions_ without interacting directly with a UI element.

Reviewer 2: I think I've said this before as well. I am not at all a gifted technical or manual writer so I don't want to second guess anybody who is. And the balance to be walked here is a fine one, I understand. Too many examples makes a messy text. This reference manual entry starts quite formal by introducing all the complicated concepts. It may make it less obvious to the more casual user that simple things are actually very simple. So they may conclude that this chapter is not what they are looking for. And maybe it isn't; maybe they should look elsewhere to see that if they simply want to link a key press to a screen button press they can do that in three/four steps:
1. press shortcut mapping button
2. hover over the button
3. press the key
(4. right click to switch off mapping mode)
Then once their use cases evolve from there, they might want to come back here to see what more is possible. I'm just worried all the complicated text here makes them believe that dt does _not_ have shortcut system that works for their needs at all.

An _action_ is usually an operation that you might undertake using darktable's point-and-click user interface.
See [common actions](#common-actions) for a list of the most common ones. For example:

-   Increase, decrease or reset sliders
-   Scroll through dropdown lists
-   Enable, expand or focus modules
-   Click buttons
-   Switch between views

A _shortcut_ is a combination of inputs that triggers one of the actions available in a given context.

Darktable comes with many predefined shortcuts using keyboard or keyboard and mouse, but you can also use various other input devices, including MIDI devices and game controllers -- see the [midi device support](../special-topics/midi-device-support.md) section for details. These are referred to as _input devices_ in this guide.

The recommended way to assign shortcuts to visual elements is the [visual shortcut mapping](#visual-shortcut-mapping) mode.

# types of actions and shortcuts

> There are two types of actions:

Reviewer 2: I understand it can be useful to introduce concepts for didactic reasons that don't exactly match reality. I myself find the concept of continuous vs discrete actions confusing (because slider adjustments are clearly not "continuous"; they are intentionally stepwise so that "toggling" up/down will hit the same cache entries so you can quickly decide which you like best between two fine-tune steps. Truly discrete adjustments would make the cache useless for this). The _real_ difference is between "directional" shortcuts (up/down scroll, move etc) and "discrete" shortcuts (a press or click, double or whatever). For keyboard+mouse use you still need to be _holding_ your last "discrete" shortcut (which _must_ include a key press) in order to be able to combine it with a "directional" one. Since you are holding it anyway, it is easy and natural to issue a series of "directional" shortcuts from it. Does that make them "continous"? With external devices you can issue "directional" shortcuts independently.

The only reason to make the distinction is that it is natural to assign "directional" shortcuts to different "default" effects than "discreet" shortcuts. So while a key press shortcut linked to a slider will show its edit pop-up, a scroll will move it up. Or down. But those are only defaults; you can explicitly override them any way you want (both assigning discrete shortcuts to an up or adown effect or assigning one of the directions to a "discrete" effect (like the popup). With the current limitation that the other direction has to be assigned to the same action (may be different element/effect) but that's mostly to guarantee that they're both available on the same screen. Could be more generalised but I doubt there's demand.

## movements in continuous shortcuts

The following movements are supported when defining continuous shortcuts:

-   Movement of the mouse scroll wheel
-   Horizontal, vertical or diagonal movement of the mouse cursor
-   Movement of a knob/joystick on an input device

When employing an input device, you can directly assign a control knob or joystick to an action. However, this will significantly reduce the flexibility of such devices, as you will be able to use the knob/joystick only to control one action. Conversely, by prefixing the movement with a key or button press you can use the same knob/joystick to control multiple actions.

## triggering multiple shortcuts at once

If both `e + scroll` and `f + scroll` are mapped to a slider (or if you have fallbacks enabled), then scrolling the mouse wheel while holding down both `e` and `f` will move both sliders.
Similarly, you could map `e double-press + scroll` and `f double-press + scroll` to two nodes in a graph (say, `yellow` and `green` in color equalizer). You can then move both sliders at once by:

> 1. pressing and releasing `e` once;

Reviewer 2: This should be a quick operation (<double click interval).

2. pressing and holding `e`;
3. pressing and holding `f`;
4. using the scroll wheel while both `e` and `f` are pressed.

This works because the repeat key counter is per-shortcut, not per key, and it is reset only after the current shortcut ends (i.e., when the repeated key is released). Thus it is possible to trigger as many actions at once as your fingers can handle, provided that all the shortcuts involved have the same number of key presses (i.e., single, double or triple presses).

## shortcuts must be unique within a view

> A single action may have multiple shortcuts but a single shortcut can only be linked to one action in a given darktable view -- you can't chain actions together except by applying a preset or style. You can, however, set up a single shortcut that does one thing in the lighttable view and another in the darkroom view.

Reviewer 2: ...except by applying a preset or style _or lua script_.

Author: I would allow users to define custom actions implemented as ordered sequences of actions. If the action is directional, it can contain any number of directional actions and the same movement is applied to all of them. This would cover probably 90% of the use cases.

Also, I would not implement a UI for that. I would keep it as a "power user" feature and only implement the logic to parse these beasts from shortcutsrc.

draft: false
---

> _Shortcuts_ are a way to control darktable using your keyboard, mouse or trackpad and other input devices. They allow you to perform _actions_ without interacting directly with a UI element.

Reviewer: ```suggestion
_Shortcuts_ are a way to control darktable using your keyboard, mouse, trackpad, or other input devices. They allow you to perform _actions_ without interacting directly with a UI element.
```

An _action_ is usually an operation that you might undertake using darktable's point-and-click user interface.
See [common actions](#common-actions) for a list of the most common ones. For example:

-   Increase, decrease or reset sliders
-   Scroll through dropdown lists
-   Enable, expand or focus modules
-   Click buttons
-   Switch between views

A _shortcut_ is a combination of inputs that triggers one of the actions available in a given context.

> Darktable comes with many predefined shortcuts using keyboard or keyboard and mouse, but you can also use various other input devices, including MIDI devices and game controllers -- see the [midi device support](../special-topics/midi-device-support.md) section for details. These are referred to as _input devices_ in this guide.

Reviewer: Please mention midi device support more near the first paragraph. This paragraph seems like a longer version of the first paragraphs. Can we merge them?

Author: Done.

The recommended way to assign shortcuts to visual elements is the [visual shortcut mapping](#visual-shortcut-mapping) mode.

# in a nutshell

If you just want to assign a specific key to toggle a button, the process is the following:

1. Click on the ![visual mapping button](./shortcuts/visual-mapping-button.png#icon) icon in the [top panel](../overview/user-interface/top-panel.md)
2. Hover the button that you want to control
3. Press the key that you want to use to control the button

Similarly, to control a slider with a shortcut you can:

1. Click on the ![visual mapping button](./shortcuts/visual-mapping-button.png#icon) icon
2. Hover the slider that you want to control
3. Press the key that you want to use to control the slider, and scroll the mouse wheel or move the mouse horizontally, vertically or diagonally

The system is very flexible and powerful, and this is just scratching the surface.
Keep on reading to understand it better and learn how to make the most of it.

> # types of actions and shortcuts

Reviewer: ```suggestion
# types of shortcuts and actions
```

There are two types of actions:

_Discrete_ actions
: E.g., executing a command, focusing a UI element or resetting a slider

_Directional actions_
: E.g., adjusting the value of a slider or scrolling through a list of presets

As the two action types are inherently different, so are the shortcuts you can use to trigger them:

- _Discrete actions_ can be triggered by _discrete shortcuts_ consisting only of key presses and optionally mouse clicks
- _Directional actions_ need also a direction (and, in the case of sliders, a magnitude). Hence, the corresponding _directional shortcuts_ will also incorporate a mouse movement

For example:

- The key `e` can be used to focus a module, or to toggle it on/off
- The key `e`, combined with up/down movements or the mouse, can be used to control the value of a slider

---

**Note:**

> By construction, a _directional shortcut_ is sufficiently expressive to trigger a _discrete action_. The opposite is not true. However, it is possible to use a combination of two _discrete shortcuts_ to control a _directional action_. For example, you can use two different shortcuts to trigger the `up` and `down` effects of a slider (see [anatomy of an action](#anatomy-of-an-action)).

Reviewer: ```suggestion
A _directional shortcut_ is sufficient to trigger a _discrete action_. The opposite is not true. However, it is possible to use a combination of two _discrete shortcuts_ to control a _directional action_. For example, you can use two different shortcuts to trigger the `up` and `down` effects of a slider (see [anatomy of an action](#anatomy-of-an-action)).
```

You can further extend a shortcut with up to three clicks of different mouse buttons. So, `e double-press + left click` or `e + right click + left click` are also valid shortcuts.

Finally, if your fingers are very nimble and you really need to go there, you can repeat the last click up to three times. So, also `e + left click + middle double-click` and `e double press + middle click + left click + right triple-click` are valid shortcuts.

Hence, the most complex shortcut can consist of 8 key/button presses:
* Three key presses of the same key, (e.g., `e`), followed by
* Three presses of the three mouse buttons, followed by
* 2 more repetitions of the last click.

Your shortcut can include one or more modifiers (`shift`, `ctrl` and `alt`). In this case, for a directional shortcut, both the last key and the modifier must be pressed down when you execute the movement. For example, `ctrl + e double-press + pan` means that while moving the mouse left and/or right both  `e` and `ctrl` must be pressed down.

---

**Note:**

> If you are a MacOs user, your shortcuts will use `cmd` instead of `ctrl` and `option` instead of `alt`.

Reviewer: ```suggestion
If you are a macOS user, your shortcuts will use `cmd` instead of `ctrl` and `option` instead of `alt`.
```

## short and long presses

By default, all key presses and button clicks in a shortcut are _short_, i.e., the key/button is pressed and immediately released. However, the last repetition of a key/button press in a shortcut can also be a _long_ one.
You can turn a normal key/button press into a long press by simply holding it a little longer before releasing it (longer than double-click speed but shorter than twice double-click speed). This applies to key presses and mouse clicks alike.
Hence `e double-press` and `e long double-press` are two distinct shortcuts that can be assigned to different actions. The associated action triggers when the key is released, which entails that a shortcut ending with a long press cannot be used for a directional shortcut.

## movements in directional shortcuts

The following movements are supported when defining directional shortcuts:

-   Movement of the mouse scroll wheel
-   Horizontal, vertical or diagonal movement of the mouse cursor
-   Movement of a knob/joystick on an input device

> When employing an input device, you can directly assign a control knob or joystick to an action. However, this will significantly reduce the flexibility of such devices, as you will be able to use the knob/joystick only to control one action. Conversely, by prefixing the movement with a key or button press you can use the same knob/joystick to control multiple actions.

Reviewer: ```suggestion
When using an input device, you can directly assign a control knob or joystick to an action. However, this will significantly reduce the flexibility of such devices, as you will be able to use the knob/joystick only to control one action. By prefixing the movement with a key or button press you can use the same knob/joystick to control multiple actions.
```

_Actions_ apply an _effect_ to an _element_ of a _widget_, where these terms are defined as follows:

widget
: Each visible part of the user interface is a _widget_. For example, the darktable application window is a widget, containing side panel widgets, each of which contains module widgets, each of which contains button, slider and dropdown list widgets etc. When assigning a shortcut to an action, you must first decide which widget it will apply to.

element
: An _element_ is the part of a UI widget that is affected by your shortcut. For example, for a slider that has a [picker](../darkroom/processing-modules/module-controls.md#pickers), you can make a shortcut activate the picker _button_ element or change the _value_ element of the slider. For a row of tabs (the row is a single widget) you can select which tab element to activate or use your mouse scroll wheel to scroll through the tabs.

effect
: A shortcut can sometimes have multiple possible _effects_ on a given _element_. For example, a button can be activated as if it was pressed with a plain mouse-click or as if it was pressed with Ctrl+click. A slider's value can be edited, increased/decreased or reset.

> So, for example, you can assign the shortcut `e + scroll` to change the value (the _effect_) of the exposure slider (the _element_) of the exposure correction _widget_.

Reviewer: ```suggestion
For example, you can assign the shortcut `e + scroll` to change the value (the _effect_) of the exposure slider (the _element_) of the exposure correction _widget_.
```

# assigning shortcuts to actions

> There are two primary methods of assigning shortcuts to actions: [visual shortcut mapping](#visual-shortcut-mapping) (recommended) and the [shortcut mapping screen](#shortcut-mapping-screen). The former makes it easier to remap multiple shortcuts at once, while the latter allows for finer control and more flexibility. Both options are detailed below.

Reviewer: ```suggestion
There are two primary methods of assigning shortcuts to actions: [visual shortcut mapping](#visual-shortcut-mapping) (recommended) and the [shortcut mapping screen](#shortcut-mapping-screen). The former makes it easier to remap multiple shortcuts at once, while the latter allows for finer control and more flexibility.
```

-   Increase, decrease or reset sliders
-   Scroll through dropdown lists
-   Enable, expand or focus modules
-   Click buttons
-   Switch between views

A _shortcut_ is a combination of inputs that triggers one of the actions available in a given context.

Darktable comes with many predefined shortcuts that use a combination of key presses and mouse movements, but you can also use various other input devices, including MIDI devices and game controllers -- see the [midi device support](../special-topics/midi-device-support.md) section for details. These are referred to as _external devices_ or just _devices_ in this guide.

# types of actions and shortcuts

There are two types of actions:

_Discrete_ actions
: E.g., executing a command, focusing a UI element or resetting a slider

_Continuous actions_
: E.g., adjusting the value of a slider or scrolling through a list of presets

As the two action types are inherently different, so are the shortcuts you can use to trigger them:

- _Discrete actions_ can be triggered by _discrete shortcuts_ consisting only of key presses and optionally mouse clicks

> - _Continuous actions_ need also a direction (and, in the case of sliders, a magnitude). Hence, the corresponding _continuous shortcuts_ will also incorporate a mouse movement

Author: Ok, in the light of this I removed the "implementation details" section and added line explaining that continuous shortcuts can also be used to trigger discrete actions.

For example:

- The key `e` can be used to focus a module, or to toggle it on/off

> - The key `e`, combined with up/down movements or your pointing device, can be used to control the value of a slider 

Author: I mention long presses on line 62, explaining that if the shortcut includes a movement the movement must be performed as the last key is held down.

My understanding is that long presses only make sense in this context, but I suspect that you will tell me that I am wrong :)

> - The key `e`, combined with up/down movements or your pointing device, can be used to control the value of a slider 

Author: Aaaaaaall right, I had gotten this completely wrong. Reintegrated. I played with it a bit, and I must confess that I find the long press very difficult to get right. If you press too long it is just ignored, and since it's not easy to get the timing right it can be a bit confusing.

If you are using keyboard and mouse, **all shortcuts must start with one or more key presses**, as mouse actions in isolation are used to navigate and interact with the UI.

If you have an external controller, you can trigger discrete actions by pressing one or more buttons on the controller. Continuous actions can use a combination of buttons and a knob/joystick movement, or just knob/joystick movement.

# anatomy of a shortcut

> Shortcuts can consist of up to three key/button presses **of the same key/button** (in a quick sequence).

Author: Thanks, very informative. Updated accordingly.

So, for example, `E`, `E + E` and `E + E + E` are all valid shortcuts to trigger discrete actions, as well as `E + left-click`, `E + right-click` or `E + E + left-click`.

> On the contrary, `E + A` is not a valid shortcut, and darktable will interpret it as a two different shortcuts: `E` followed by `A`.

Reviewer 2: ```
F1;double=iop/colorequal/brightness/red
F1;triple=iop/toneequal/blacks
F2;double=iop/colorequal/brightness/orange
F2;triple=iop/toneequal/deep shadows
F3;double=iop/colorequal/brightness/yellow
F3;triple=iop/toneequal/shadows
F4;double=iop/colorequal/brightness/green
F4;triple=iop/toneequal/light shadows
F5;double=iop/colorequal/brightness/cyan
F5;triple=iop/toneequal/mid-tones
F6;double=iop/colorequal/brightness/blue
F6;triple=iop/toneequal/dark highlights
F7;double=iop/colorequal/brightness/lavender
F7;triple=iop/toneequal/highlights
F8;double=iop/colorequal/brightness/magenta
F8;triple=iop/toneequal/whites
```
You're probably overthinking this. Merge the above and `F1+F1+F2+F3+scroll` will move three color equalizer nodes (if you have fallbacks enabled) and `F3+F3+F3+F2+F1+scroll` will move three tone equalizer nodes.

> On the contrary, `E + A` is not a valid shortcut, and darktable will interpret it as a two different shortcuts: `E` followed by `A`.

Author: Ok, thanks for the snippet, I tried it and yes, I can move the three nodes together. I still do not understand the logic, though.

As exemplified above, `iop/colorequal/brightness/orange` is triggered by `F2,double`, which I parse as "double press of F2".

In the key combination `F1 + F1 + F2` there is no double press of `F2`, so why should `iop/colorequal/brightness/orange` trigger?

I have seen that it does. The question is why? It shouldn't according to (how I parse) the specification of the shortcut.

Your shortcut can include one or more modifiers (`Shift`, `Ctrl/Cmd` and `Alt/Option`). In this case, the modifier(s) have to be held down while executing the remainder of the shortcut. So, `Ctrl + E + E` means holding down `Ctrl` while pressing `E` twice in a rapid sequence.

If you are defining a continuous shortcut, then the movement part of the shortcut must be executed while the last key, mouse or controller button is held down.

For example, the shortcut `E + E + pan`, can be activated by pressing `E` twice, holding down `E` on the second press and moving the mouse horizontally. While you are holding down `E`, moving the mouse horizontally will adjust the value associated with the shortcut's action. As you release the `E` key, moving the mouse horizontally will just move the mouse on the screen.

**Shortcuts must be unique within a view.**
A single action may have multiple shortcuts but a single shortcut can only be linked to one action in a given darktable view -- you can't chain actions together except by applying a preset or style. You can, however, set up a single shortcut that does one thing in the lighttable view, say, and another in the darkroom view.

> **Additional modifiers.** As mentioned above, the only valid modifiers are the Shift, Ctrl/Cmd and Alt/Option keys on the keyboard. You can define additional keys (or device buttons) as modifiers by assigning keys/buttons to the "global/modifier" action. However, these will merely function as extra Ctrl, Alt or Shift keys -- you cannot create "new" modifiers.

Reviewer 2: Are Cmd and Option for Apple users mentioned everywhere in the docs where Ctrl and Alt are used? Or should/is this be something that the introduction mentions once. It could be somewhat confusing for everyone else.

# anatomy of a shortcut

The simplest shortcut consists of just one key press (e.g., `E`).
The same key can be repeated up to three times, so also `E + E` and `E + E + E` are valid, and distinct, shortcuts.
You can further extend a shortcut with up to three clicks of the same mouse button. So, `E + E + left-click` or `E + right-click + right-click` are also valid shortcuts.

> Finally, if your fingers are very nimble and you really need to go there, you can add either or both of the mouse buttons that you haven't incorporated yet. So, also `E + left-click + middle-click` and `E + E + right-click + right-click + right-click + left-click + middle-click` are valid shortcuts.

Reviewer 2: Aaah! Now with more testing to remind me, the "triple click" of the mouse button has to be the _final_ button. So it would have to be `E+E+E+right+middle+left+left+left`.

Author: Oh well, this makes much more sense :) Fixed.

Hence, the most complex shortcut can consist of 8 key/button presses:
* Three key presses of the same `key`, followed by
* Three presses of any mouse button, followed by
* 1 press of each remaining mouse button.

> Your shortcut can include one or more modifiers (`Shift`, `Ctrl` and `Alt`). In this case, the modifier(s) have to be held down while executing the remainder of the shortcut. So, `Ctrl + E + E` means holding down `Ctrl` while pressing `E` twice in a rapid sequence.

Reviewer 2: I'm not suggesting any wording here; just trying to help improve _your_ mental modal and then you can decide if it is worth explaining to the wider public.

You _don't_ have to press the modifiers before pressing the shortcut keys. You can add them later. So for example if triple-press A is assigned to a slider (and lets assume you have fallbacks enabled). You can do A+A+A hold, then scroll for "normal sized" adjustments. Then (while still holding the A) add Ctrl for some final fine-tuning. Personally I prefer mouse  moves, so I first scroll and then make some final left-right moves at *.1.

Reviewer 2: Shortcuts don't get triggered when you _press_ a key (or button); they get triggered when you _release_ a key (ok, with the exception of _hold_ shortcuts...). Modifiers don't trigger shortcuts, they only "modify" them. So if you press e, then Ctrl then release e, you should get the Ctrl+e shortcut. If you do it _fast_ enough, otherwise it will be a long press or ignored. 

So it has nothing to do with fallbacks but it is more relevant for what you call continuous shortcuts. But also if you combine keypresses with mouse clicks. Ctrl+e+left click you can add the Ctrl _after_ pressing (and holding) the e (_before_ or even _during_ the click).

Disclaimer not behind keyboard so not testing/verifying right now.

-   ![Don't signal](./shortcuts/no-signal.png#icon) indicates that there is no mappable widget under the cursor.

**To define new shortcuts:**

> press a key combination while hovering over a mappable widget. A _default action_ will be assigned to that shortcut based on the type of widget and whether your shortcut includes a movement. See the [common actions](#common-actions) section for examples of some of the defaults.

Reviewer 2: "whether your shortcut includes a movement" you mean whether it is a continuous shortcut?

Author: Done

-   ![Don't signal](./shortcuts/no-signal.png#icon) indicates that there is no mappable widget under the cursor.

**To define new shortcuts:**

> press a key combination while hovering over a mappable widget. A _default action_ will be assigned to that shortcut based on the type of widget and whether it is directional. See the [common actions](#common-actions) section for examples of some of the defaults.

Reviewer: ```suggestion
Press a key combination while hovering over a mappable widget. A _default action_ will be assigned to that shortcut based on the type of widget and whether it is directional. See the [common actions](#common-actions) section for examples of some of the defaults.
```

You can assign as many shortcuts as you like in a single mapping session and then exit mapping mode when you are finished by clicking the ![visual mapping button](./shortcuts/visual-mapping-button.png#icon) icon again or right-clicking anywhere on the screen.

**To explore already defined shortcuts:**

> left-click on a mappable widget to open the shortcut mapping screen for that widget (see below). Left-click anywhere else on the screen to open the shortcut mapping screen, expanded (where possible) based on the part of the screen you have clicked on. This screen can be used to alter the action assigned to a shortcut and to configure shortcuts for non-visual actions. Entering the shortcut mapping screen exits visual shortcut mapping mode.

Reviewer: ```suggestion
Left-click on a mappable widget to open the shortcut mapping screen for that widget (see below). Left-click anywhere else on the screen to open the shortcut mapping screen, expanded (where possible) based on the part of the screen you have clicked on. This screen can be used to alter the action assigned to a shortcut and to configure shortcuts for non-visual actions. Entering the shortcut mapping screen exits visual shortcut mapping mode.
```

**To update existing shortcuts:**

> define a second shortcut against the same widget. If you attempt to reallocate an existing shortcut to a new action, you will be notified of the conflict and prompted to replace the existing shortcut.

Reviewer: ```suggestion
Define a second shortcut against the same widget. If you attempt to reallocate an existing shortcut to a new action, you will be notified of the conflict and prompted to replace the existing shortcut.
```

**To change the default speed of a slider:**

> scroll with your mouse wheel while in visual mapping mode (without pressing any other buttons/keys) when hovering over the slider -- scroll up to increase and down to decrease. When you leave mapping mode, normal mouse scrolls over that slider will change its value with the adjusted speed.

Reviewer: ```suggestion
Scroll with your mouse wheel while in visual mapping mode (without pressing any other buttons/keys) when hovering over the slider -- scroll up to increase and down to decrease. When you leave mapping mode, normal mouse scrolls over that slider will change its value with the adjusted speed.
```

## shortcut mapping screen

The shortcut mapping screen can be accessed from the global preferences dialog or by right-clicking on the visual mapping button. This screen allows access to all available actions, including some that are not directly linked to a UI widget.

The top panel of the shortcut mapping screen shows a list of available UI widgets/actions and the bottom panel shows the shortcuts currently assigned to them. Fields that can be changed by user action are shown in bold.

---

**Note:**

> You can **search** the top and bottom panels using the text entry boxes at the bottom of the screen, and use the **up/down** arrow keys to navigate between matches. 

Reviewer: ```suggestion
You can search the top and bottom panels using the text entry boxes at the bottom of the screen, and use the up/down arrow keys to navigate between matches. 
```

# common actions

The following is a list of some of the actions to which you can assign shortcuts, organized by widget type. This is not an exhaustive list and you are encouraged to browse the shortcut mapping screen for a complete list of available actions.

---

**Note:**
some actions may have no effect. For example, all sliders include a _button_ element, regardless of whether such a button is actually present alongside a given slider.

---

## global actions

Actions in the "global" section of the shortcut mapping screen can be executed from any darktable view. Most of these actions do not have specific _elements_ as they are used to perform one-off operations.

For example, the predefined shortcut `tab` triggers the action `globals/panels/all`, which toggles the visibility of all the side panels in the current view.

## view-specific actions

Actions in the "views" section can only be executed from the specified darktable view. As with global actions, most do not have specific _elements_ as they are used to perform one-off operations.

> For example, the predefined shortcut `ctrl + b` triggers the action `views/darktable/guide lines/toggle`, which toggles guide lines in the darktable view.

Author: Added

# common actions

The following is a list of some of the actions to which you can assign shortcuts, organized by widget type. This is not an exhaustive list and you are encouraged to browse the shortcut mapping screen for a complete list of available actions.

---

**Note:**

> some actions may have no effect. For example, all sliders include a _button_ element, regardless of whether such a button is actually present alongside a given slider.

Reviewer: ```suggestion
Some actions may have no effect. For example, all sliders include a _button_ element, regardless of whether such a button is actually present alongside a given slider.
```

: Acts as a _button_ that resets all module parameters when activated. The _ctrl-activate_ action can be used to re-apply any automatic presets for that module.

_presets_
: Allows you to select actions from the [presets](../darkroom/processing-modules/presets.md) menu (e.g. edit, update, previous, next). The default action, when assigning a discrete shortcut to a _preset_ element, is to display a list of the available presets for selection.

When a shortcut is assigned to a utility module, the default action is to _toggle_ the _show_ element (expand/collapse the module).

### processing modules

Additionally, processing modules have the following elements:

_enable_
: Acts as a _toggle_ that switches the module on and off.

_focus_

> : Acts as a _toggle_ that focuses or de-focuses the module. This is useful for modules such as [_crop_](../module-reference/processing-modules/crop.md) or [_tone equalizer_](../module-reference/processing-modules/tone-equalizer.md), whose on-screen controls are only activated when those modules have focus. For _crop_, changes are saved only when the module loses focus.

Reviewer: ```suggestion
: Acts as a _toggle_ that focuses or unfocuses the module. This is useful for modules such as [_crop_](../module-reference/processing-modules/crop.md) or [_tone equalizer_](../module-reference/processing-modules/tone-equalizer.md), whose on-screen controls are only activated when those modules have focus. For _crop_, changes are saved only when the module loses focus.
```

You can alter the value of a slider more quickly or slowly than normal by defining the _speed_ of the action in the shortcut mapping screen. By default a _value_ (or _force_) effect is given a speed of 1.0, which means that it is changed at the default rate defined by the given slider. You can alter the slider more quickly by increasing the speed (a speed of 10 makes the action 10x faster) or more slowly by decreasing it (a speed of 0.1 makes the action 10x slower).

# fallbacks (a.k.a. generalized shortcut extensions)

> As discussed above, a _widget_ can have multiple different _actions_ applied to it, and you may want to set shortcuts to control several of them. For example, to toggle a module on/off, expand/collapse it, and reset it to its default state. Normally, this would require setting 3 shortcuts for each module. If you want to control 5 modules in this fashion, you have to define 15 shortcuts. Setting up all these shortcuts one by one can be tedious and error prone.

Reviewer: A _widget_ can have multiple different _actions_ applied to it, and you may want to set shortcuts to control several of them. For example, to toggle a module on/off, expand/collapse it, and reset it to its default state. Normally, this would require setting 3 shortcuts for each module. If you want to control 5 modules in this fashion, you have to define 15 shortcuts. Setting up all these shortcuts one by one can be tedious and error prone.
```

Author: The comment is just a repetition of the suggested edit, isn't it?
